src/main/java/ClientConnection.java

```diff
@@ -16,7 +16,7 @@ public ClientConnection(int portNumber) throws IOException {
     server = new ServerSocket(portNumber);
   }
 
-  public String receiveRequest() throws IOException {
+  public String[] receiveRequest() throws IOException {
```

Comment:
I'm a little concerned about this return type, for a couple reasons:

1) Your `ClientConnection` is currently not tested. To what degree it should be tested, if at all, is debatable, but if it does remain not tested at all, or minimally tested, it should have as little logic around it as possible. IMO the `split` on whitespace below qualifies as enough additional logic that it should definitely be tested.

2) The `String[]` return type and splitting on whitespace demonstrates that knowledge about the request format is leaking into this `ClientConnection` class and away from some other object that should be responsible for understanding what constitutes an HTTP request. To put that another way: when this returned a `String`, it's only responsibility was to collect data from an input stream; it was something else's responsibility to understand HTTP and attempt to coerce that data into an HTTP request. Now we have this object splitting apart components of the HTTP request _and_ passing those components via array to a second object that now implicitly depends on this first object (to put the pieces of the request in the proper order).

3) `String[]` introduces the need to do an annoying check for `IndexOutOfBoundsException`. Blegh.